### ISSUE TITLE
Extensibility for fieldtypes

### ISSUE THREAD

**Issue by timkelty** [boboldehampsink/import#77]
I was wondering if there was a plan to support an architecture for fieldtypes that store data in other tables than `craft_content`.

For instance, I have a bunch of map data I would love to be able to import into a Smart Map field, which stores things in it's own table. Would be great if there was an easy way to write a service to handle this.

**Comment by boboldehampsink** [comment]
This extensibility has been added already as of late, by the creator of SmartMap ;-) cc @lindseydiloreto
He can probably tell you more about it

**Comment by timkelty** [comment]
oh nice! Thanks!

**Comment by lindseydiloreto** [comment]
There are two hooks you can use for that purpose...

**registerImportOptionPaths** - Prep the field mapping _before_ the import is run. This allows you to customize the `<select>` menu option(s) for your field. Smart Map uses it to split the standard _one_ option into a _group_ of options for all of the subfields (street1, street2, city, state, zip, country, lat, lng).

**modifyImportRow** - Manipulate the data _as the import is running_. You'll be able to tweak your data as it goes into the database.

I'm going to break these out into separate comments, so I can explain them in more detail. :)

**Comment by lindseydiloreto** [comment]
## registerImportOptionPaths

>_Disclaimer: Not all custom field types will need this hook. If your field type does not need to customize the standard `<select>` menu, then skip ahead to the `modifyImportRow` hook._

When you are mapping your CSV columns to Craft fields, Import let's you choose which field you want to assign your data to. There is a `<select>` menu with a list of Craft fields, and you select the destination for each column. By default, this is a 1:1 mapping.

This was a problem for Smart Map, however, because an "Address" field is a single field, yet it's capable of consuming up to _eight pieces of data_. And that data was expected to be imported as completely separate CSV columns.

So the purpose of the `registerImportOptionPaths` hook is to give you the opportunity to **replace the standard `<option>` tag with something custom**. In essence, you are providing a _template_ which contains a custom `<option>` (or multiple options) to handle the data mapping in a way that is more preferable for your field type.

Put this in your main `MyPlugin.php` file, and edit accordingly:

```php
public function registerImportOptionPaths()
{
	return array(
		'MyPlugin_MyFieldType' => 'myplugin/_customImportOption.html',
	);
}
```

You are being given the opportunity to say **"for fields of this type, use my custom template in place of the default `<option>` tag".** Make sure your `_customImportOption.html` template outputs one or more `<option>` tags... They will be injected into the field select dropdown menu.

Here's the final output for Smart Map:

![options](https://cloud.githubusercontent.com/assets/5309692/10528359/9721b560-734a-11e5-91fe-f7e5f3aa2529.png)

The red box shows the result of our `registerImportOptionPaths` hook... Each CSV column can now be mapped to a different _subfield_ of the Address field. Keep in mind, **this is one field, generating eight `<option>` tags.**

In summary:
 1. Create a short template to replace the default `<option>`
 2. Map that template to your field type with this hook

If you want to dig in the code a little more and see how it works in even greater detail, [check out this chunk of code here...](https://github.com/boboldehampsink/import/blob/00818471336c92d9461e174096a14f3d30b68697/templates/types/entry/_map.twig#L50-L58)

**Comment by lindseydiloreto** [comment]
## modifyImportRow

This one is a little easier to explain... It simply allows you to make any last-minute adjustments to your data before it gets stored in the database.

Something like this is a good starting point:

```php
/**
 * Modify data just before importing
 *
 * @param BaseElementModel $element  The current element receiving import data.
 * @param array            $map      Mapping of data between CSV -> Craft fields.
 * @param array            $data     Raw data provided for this row.
 */
public function modifyImportRow($element, $map, $data)
{
	// Map data to fields
	$fields = array_combine($map, $data);

	// Initialize content array
	$content = array();

	// Arrange your content in a way that makes sense for your plugin
	foreach ($fields as $handle => $value) {
		$content[$handle] = $value;
	}

	// Set modified content
	$element->setContentFromPost($content);
}
```

The `foreach` loop is probably where you'll do the most customizing. However, feel free to edit any (or all) of this method to achieve your desired results.

>_Full disclosure: I did not write this hook, it was originally authored by @freddietilley. He may have more to add, or even correct me if I'm mistaken on any of this._

**Comment by lindseydiloreto** [comment]
Of course, now that I've posted this massive breakdown of these two hooks, I realize that you're actually trying to import via Smart Map (and not a custom plugin). The good news is, **this has already been added to Smart Map**... I'll send you a beta copy which includes this functionality.

Hopefully this extensive breakdown of "how to use these hooks" turns out to be helpful to any other plugin developers in the future. :)

**Comment by timkelty** [comment]
Thanks! I have other fields that _are_ custom plugins, so your post is very helpful.

**Comment by lindseydiloreto** [comment]
Great! :thumbsup:
